sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -6129,6 +6129,20 @@ pi_result piextUSMEnqueueMemset(pi_queue Queue, void *Ptr, pi_int32 Value,
       Count, NumEventsInWaitlist, EventsWaitlist, Event);
 }
 
+// Helper function to check if a pointer is a host pointer.
+static bool IsHostPointer(pi_context Context, const void *Ptr) {
+  // Query the device of the allocation
+  ze_device_handle_t ZeDeviceHandle;
+  ze_memory_allocation_properties_t ZeMemoryAllocationProperties = {};
+
+  // Query memory type of the pointer
+  ZE_CALL(zeMemGetAllocProperties,
+          (Context->ZeContext, Ptr, &ZeMemoryAllocationProperties,
+           &ZeDeviceHandle));
+
+  return (ZeMemoryAllocationProperties.type == ZE_MEMORY_TYPE_HOST);
```

Comment:
This will only return true for allocations that were created with `ZeMemAllocHost`. Don't we want to use copy engine for shared allocations as well? Also, for pointers created with host malloc() I think we should use copy engine for them too (I expect that in this case allocation type reported by the driver is `ZE_MEMORY_TYPE_UNKNOWN`.

Reply:
Good point. I think I can change this function to 'IsDeviceMemory' and modify the check to !IsDeviceMemory(Src) && !IsDeviceMemory(Dst).

Reply:
Sounds good. And please add a comment why D2D copy is not using copy engine.

Reply:
Uploaded changes. Please review. Thanks.